### ISSUE TITLE
UTF-8 encoding when dumping data for external graders

### ISSUE THREAD

**Issue by lumetta** [PrairieLearn/PrairieLearn#3183]
I used the PL Python autograder as a base for writing a C grader and ran across an issue that I think I inherited from your code (and with which I'm not familiar enough with Python to fix on my side).

Specifically, the Docker sets an environment variable (PYTHONIOENCODING) to force Python to default to UTF-8 encodings for I/O.  (I tried turning it off in my Docker ... caused some other more serious failures, it seems.)

However, when PL writes the data dict to a file, it doesn't use UTF-8 encoding.  I would have given you an example using one of the demo problems, but your examples don't seem to include any strings, whereas my first example was a fill-in-the-blank problem based on strings (ssltest/S17-MT1-P6 in ECE220).  If one puts, for example, a fancy apostrophe (such as "printf( "Student %d’s score:", student )") in a string, the data.json file is not UTF-8 encoded, so the JSON load code (in my pl_main.py, and I think it was copied from yours, but I may have moved it from another file--didn't change it, though, so it's in Python autograder, too) throws an exception.

--Steve

**Comment by mwest1066** [comment]
We write `data.json` using [fs-extra](https://www.npmjs.com/package/fs-extra) here: https://github.com/PrairieLearn/PrairieLearn/blob/423554a832aac30182c02fbe6b87a4cb3ecc89e2/lib/externalGraderCommon.js#L148 which I believe defaults to `utf8`: https://nodejs.org/api/fs.html#fs_fs_writefile_file_data_options_callback

We read it here: https://github.com/PrairieLearn/PrairieLearn/blob/423554a832aac30182c02fbe6b87a4cb3ecc89e2/graders/python/python_autograder/pl_execute.py#L48

@lumetta have you confirmed that the file on disk is not being written as UTF-8? Can you reproduce this locally? Is there a small change to an exampleCourse question that will show this problem? I'm not quite clear on where you are putting the `printf()` statement.

Maybe you are referencing `results.json` rather than `data.json`? I don't see `data.json` referenced in `pl_main.py`. I do see one `open()` call that's missing an `encoding` here: https://github.com/PrairieLearn/PrairieLearn/blob/423554a832aac30182c02fbe6b87a4cb3ecc89e2/graders/python/python_autograder/pl_main.py#L141 compared to the non-error path here: https://github.com/PrairieLearn/PrairieLearn/blob/423554a832aac30182c02fbe6b87a4cb3ecc89e2/graders/python/python_autograder/pl_main.py#L132

**Comment by lumetta** [comment]
Sorry, my fault.  The code in question is from your pl_execute.py.   I don't use the Python autograder and didn't update my copy before borrowing the code as a starting point.  The 'utf-8' encoding on the open seems to fix the problem.  Without it, Python JSON requires ASCII, not just UTF-8, so only 7-bit values are allowed.  Fancy quotes are coded as extended ASCII and break json.load.  I examined the files within Docker launched from PL to see the sequence of offending characters (python mentioned the leading xE2 byte).

I did a workaround and will share it if you think it'd be useful (I'd do a pull, but obviously my Python skills are lacking).  I read the file as a binary file (byte sequence), replaced the fancy quotes, and looked for other non-ASCII that I hadn't thought of.  Sent back a format error complaining about non-ASCII submissions in those cases to avoid student loss of attempt.  Works now (but pretty hackish).

I'm going to leave my workaround for my C grading in place because we need to diff output, and I don't think diff will identify quotes with fancy quotes.  gcc fails, too (maybe there's an option to allow).

Sorry for the confusion.

(I'd have given you an example earlier, but I'd have to write a new Python question with a pl-string-input to do it.  I used pl-string-input for each blank in the C code.  I looked at the examples and found only code entry, which saves as code.  Interestingly, there's some translation of quotes in both python and bash under Cygwin, but not in vi--if I cut and paste fancy quotes, they are turned into ASCII quotes in python and bash.  Not sure what code is doing that.  We did have some browsers producing the fancy quotes, it seemed (in most cases, students copied a fancy quote from the problem statement). )

--Steve
